core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/PhysicalServices.java

```diff
@@ -824,60 +819,76 @@ public void updateInternalPhysicalPaths(DDiagram diagram) {
-    // Internal edges representing the physical path
-    HashMap<PhysicalPath, Set<DEdge>> physicalPathToEdgesMap = computePhysicalPathToEdgesMap(diagram);
-
     // Find displayed physical paths
     HashMap<PhysicalPath, DNode> displayedPaths = computePhysicalPathToNodeMap(diagram);
-
+    HashMapSet<DSemanticDecorator, RGBValues> viewColors = new HashMapSet<>();
     // Find displayed Physical Links
     HashMap<PhysicalLink, DEdge> displayedPhysicalLinks = computePhysicalLinkToEdgeMap(diagram);
-
     // Find physical links that must be colored
     HashMap<DEdge, Set<PhysicalPath>> coloredLinks = computeEdgeToPhysicalPathsMap(displayedPaths,
         displayedPhysicalLinks);
-
-    ArrayList<DEdge> displayedLinks = new ArrayList<DEdge>();
+    ArrayList<DEdge> displayedLinks = new ArrayList<>();
     
     for (Entry<PhysicalPath, DNode> entry : displayedPaths.entrySet()) {
       DNode node = entry.getValue();
-      PhysicalPath physicalPath = entry.getKey();
-
+      PhysicalPath path = entry.getKey();
       updatePhysicalPathNodeColor(node, displayedPaths.values());
-      RGBValues physicalPathColor = ShapeUtil.getNodeColorStyle(node);
-
-      // Customize physical paths
-      for (PhysicalLink link : PhysicalPathExt.getFlatPhysicalLinks(physicalPath)) {
-        if (displayedPhysicalLinks.containsKey(link)) {
-          DEdge currentEdge = displayedPhysicalLinks.get(link);
-          RGBValues color = (coloredLinks.get(currentEdge).size() == 1) ? physicalPathColor : ShapeUtil.getBlackColor();
-          customizePhysicalLinkEdgeStyle(currentEdge, color);
+      RGBValues color = ShapeUtil.getNodeColorStyle(node);
+      PhysicalPathInvolvementGraph graph = PhysicalPathCache.getInstance().getInvolvementGraph(path);
+
+      // Add color for related physical links
+      for (InvolvementEdge edge : graph.getEdges().values()) {
+        PhysicalLink involvedPL = graph.getInvolvedPhysicalLink(edge);
+        if (involvedPL != null) {
+          for (DSemanticDecorator view : DiagramServices.getDiagramServices().getDiagramElements(diagram, involvedPL)) {
+            viewColors.put(view, color);
+            if (view instanceof DEdge) {
+              Set<PhysicalPath> paths = coloredLinks.computeIfAbsent((DEdge) view, k -> new HashSet<>());
+              paths.add(path);
+            }
+          }
         }
       }
       
-      if (physicalPathToEdgesMap.containsKey(physicalPath)) {
-        displayedLinks.addAll(physicalPathToEdgesMap.get(physicalPath));
+      // Create and highlight internal links
+      PhysicalPathInternalLinksGraph linksGraph = PhysicalPathCache.getInstance().getInternalLinksGraph(graph);
+      for (InternalLinkEdge edge : linksGraph.getEdges().values()) {
+        PhysicalPort source = edge.getSource().getSemantic();
+        PhysicalPort target = edge.getTarget().getSemantic();
+        DDiagramElement sourceNode = DiagramServices.getDiagramServices().getDiagramElement(diagram, source);
```

Comment:
On multipart, a port might be displayed several times, maybe a 
for (sourceNodes
  for(targetNodes 
     createInternalLink
     
in createInternalLink there is a check if valid (same container node) i think